### PR TITLE
Lps 164449

### PR DESCRIPTION
**Pull request by thienquach87**
Hi team,

This is the PR for fixing the [LPS-164449](https://issues.liferay.com/browse/LPS-164449), the original has passed Object team's review at https://github.com/liferay-objects/liferay-portal/pull/1220. There is a part of code relating to properties for Oracle ,so could you help me to review and give feedback.

**Update**:
Currently, in case of using HSQL Database, if the createdDate and modifiedDate are used to filter records, there is an exception thrown "invalid datetime...". The reason is because the current date format used is not supported in HSQL. To fix this issue I have solution as below:
**Solution:**
- Setting sql.syntax_ora=true to enable support for non-standard syntax of Oracle by adding jdbc.default.url=jdbc:hsqldb:${liferay.home}/data/hypersonic/lportal;sql.syntax_ora=true to portal.properties file.
- Because HSQL only support four date's patterns:
   1  pattern = "DD-MON-YY"
   2  pattern = "DD-MON-YYYY"
   3  pattern = "DD-MON-YYYY HH24:MI:SS"
   4  pattern = "DD-MON-YYYY HH24:MI:SS.FF"
,so I implemented source to convert the format date to "DD-MON-YYYY HH24:MI:SS.FF" pattern supported by HSQL to fix this issue.

Thanks

### CODE REVIEW

**Comment by dantewang** [comment]
@thienquach87 Code conflicts, please rebase, thanks!

**Comment by thienquach87** [comment]
ci:test:relevant

**Comment by thienquach87** [comment]
> @thienquach87 Code conflicts, please rebase, thanks!

I updated source to fix conflict source.

**Comment by thienquach87** [comment]
@gabrielwas,
The code in object part is edited a lot, so I updated source to apply my logic. Could you help to take a look on my changes.

**Comment by Kvale1733** [comment]
@thienquach87 No regressions found.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java [hunk 1]
```diff
@@ -343,12 +387,22 @@ else if (Objects.equals(BinaryExpression.Operation.NE, operation)) {
 	}
 
 	private Object _getValue(Object left, Object right) {
+		EntityField entityField = _getEntityField(left);
+
+		EntityField.Type entityType = entityField.getType();
+
+		if (entityType.equals(EntityField.Type.DATE_TIME) &&
+			StringUtil.startsWith(_getDriverName(), "HSQL")) {
+
```
Comment (yuhai) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java:396]: @thienquach87, you may simplify the code fragment and replace StringUtil.startsWith() with the following code
```
DB db = DBManagerUtil.getDB();
if (db.getDBType() == DBType.HYPERSONIC) {}
```

File: portal-impl/src/com/liferay/portal/tools/HypersonicLoader.java [hunk 2]
```diff
@@ -95,10 +95,13 @@ public HypersonicLoader(
...
+			";hsqldb.write_delay=false;shutdown=true;sql.syntax_ora=true";
+
 		try (Connection connection = DriverManager.getConnection(
 				StringBundler.concat(
 					"jdbc:hsqldb:", sqlDir, "/", databaseName,
-					";hsqldb.write_delay=false;shutdown=true"),
+					connectionProperties),
 				userName, password)) {
 
 			if (Validator.isNull(fileNames)) {
```
Comment (yuhai) [inline portal-impl/src/com/liferay/portal/tools/HypersonicLoader.java:107]: @thienquach87, the fix looks unrelated with the second commit. Since HypersonicLoader class isn't loaded during portal runtime. If the fix can take effect, why not add the fix to the property **jdbc.default.url** in portal.properties?

**Comment by catarineaguiar** [comment]
@thienquach87 if you make changes, re-run relevant and remove the QA label, please. Thanks.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java (hunk 1 above)
Comment (thienquach87) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/odata/filter/expression/PredicateExpressionVisitorImpl.java:378]: thank you, I updated source based on your suggestion

File: portal-impl/src/com/liferay/portal/tools/HypersonicLoader.java (hunk 2 above)
Comment (thienquach87) [reply portal-impl/src/com/liferay/portal/tools/HypersonicLoader.java]: You are right @yuhai, we also add jdbc.default.url=jdbc:hsqldb:${liferay.home}/data/hypersonic/lportal;sql.syntax_ora=true to add the sql.syntax_ora property to 'true' value. However if we do it in this way, it is necessary to take note it to end user.
Finally, I updated source following your comment. Please help review it again and leave your feed-back. 

Thank you!

**Comment by thienquach87** [comment]
ci:test:sf

**Comment by thienquach87** [comment]
ci:test:relevant

**Comment by yuhai** [comment]
@thienquach87, since the current fix only is in the Object component, could you please send the fix to liferay-objects user to review?

In addition, I want to mention one thing if we want to add to the property sql.syntax_ora=true.
As the statement from https://hsqldb.org/doc/2.0/guide/compatibility-chapt.html

> Use SET DATABASE SQL SYNTAX ORA TRUE or the equivalent URL property sql.syntax_ora=true to enable support for some non-standard syntax of Oracle.

In the expected fix, when liferay uses non-standard syntax of Oracle, and it won't work in HSQL database. After we add the property, the issue can be resolved. That is to say, it should only add this property as the fix.  Besides that, we need to firstly add one test to expose the issue if possible. 

I will firstly close the pr.

Thanks,
Hai

**Comment by thienquach87** [comment]
@yuhai 
Thank you for your supports
